### PR TITLE
app-state tab

### PR DESCRIPTION
**Pull request by daiyi**
Work in progress!

This tab shows the app state. Lots of interesting future features to be added, but I'm putting up a PR for comments and discussion.

Currently it's statically showing the app-state, aka the atom `db/app-db`!

I imagine some of the work here would be useful/reusable for #52.

preview:

```clojure
{:todos
 {1 {:id 1 :title "eat snacks" :done false}
  2 {:id 2 :title "bike to lake" :done false}}
 :showing :all}
```

WIP app-state tab

### CODE REVIEW

**Comment by Dexterminator** [comment]
The pretty printing doesn't quite seem to work for me:

File: src/day8/re_frame/trace/app_state.cljs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    (nil? obj)       "nil"
+    (keyword? obj)   "keyword"
+    (symbol? obj)    "symbol"
+    :else (pr-str (type obj))))
+
+(defn view
+  [data]
+  (if (coll? data)
+    [:div  {:class (str (namespace-css "collection") " " (namespace-css (css-munge (type-string data))))}]
+    [:span {:class (str (namespace-css "primative") " " (namespace-css (css-munge (type-string data))))} (str data)]))
```
Comment (danielcompton) [inline src/day8/re_frame/trace/app_state.cljs]: primitive

File: src/day8/re_frame/trace/app_state.cljs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+(defn crawl
+  [data]
+  (if (coll? data)
+    (into (view data) (mapv crawl data))
```
Comment (danielcompton) [inline src/day8/re_frame/trace/app_state.cljs]: This recursion will blow the stack eventually, you'll need to switch to loop/recur at some point.

File: resources/day8/re_frame/trace/main.less [hunk 3]
```diff
@@ -321,5 +329,55 @@
...
+  .re-frame-trace--nil {
+    text-decoration: line-through;
+    font-style: italic;
+  }
+  .re-frame-trace--keyword {
+    color: @light-purple;
+  }
+  .re-frame-trace--symbol {
+  }
+  .re-frame-trace--cljs-core-PersistentArrayMap,
```
Comment (danielcompton) [inline resources/day8/re_frame/trace/main.less]: You don't want to be adding classes for concrete datatypes, you can check for `vector?`, `map?` instead. You could use the interface name if you liked though (IPersistentVector).

File: src/day8/re_frame/trace/app_state.cljs [hunk 4]
```diff
@@ -0,0 +1,42 @@
+(ns day8.re-frame.trace.app-state
+  (:require [reagent.core :as r]
+            [clojure.string :as str]
+
+            [cljs.pprint :refer [pprint]]))
+
+
+(defn css-munge
```
Comment (danielcompton) [inline src/day8/re_frame/trace/app_state.cljs]: It's good to split these functions out. I'd recommend making a short function name which combines them both so you can use it everywhere consistently, and make it easier to see what's going on.

**Comment by danielcompton** [comment]
@daiyi just wanted to mention before you go too far down this path, that a pretty printing approach, or anything which prints the whole app-db in one go will be too slow to be merged. Wasn't sure if I'd got that across last time we talked, but don't want you to spend too much time on the non devtools formatter version.

**Comment by daiyi** [comment]
Yep, I know we're going to do the formatters now! I was keeping this PR open as a bookmark, but we can close it if you'd like less clutter

**Comment by danielcompton** [comment]
You can leave it open if it's useful for you, I'm not too bothered either way. I thought I saw some recent activity on this which I was why I quickly dashed the note off above, but I think I must have been mistaken.

**Pull request closed without merging**